code/modules/mob/mob_layering.dm

```diff
@@ -0,0 +1,102 @@
+/mob/proc/get_base_layer()
+	if(lying)
+		return LYING_MOB_LAYER
+	return initial(layer)
+
+/mob/reset_layer()
+	var/last_layer = layer
+	var/new_layer = get_base_layer()
+	if(isturf(loc))
+		if(buckled && buckled.buckle_layer_above)
+			new_layer = buckled.layer + ((buckled.dir == SOUTH) ? -0.1 : 0.1)
+		else if(length(grabbed_by))
+			var/draw_under = TRUE
+			var/adjust_layer = FALSE
+			for(var/obj/item/grab/G AS_ANYTHING in grabbed_by)
+				if(get_dir(G.assailant, src) & SOUTH)
+					draw_under = FALSE
+				if(G.current_grab.adjust_plane)
+					adjust_layer = TRUE
+			if(adjust_layer)
+				new_layer = layer + (draw_under ? -0.01 : 0.01)
+	if(new_layer != last_layer)
+		layer = new_layer
+		UPDATE_OO_IF_PRESENT
+
+/mob/reset_plane()
+	var/last_plane = plane
+	..()
+	var/new_plane = plane
+	if(isturf(loc))
+		if(buckled && buckled.buckle_layer_above)
+			new_plane = buckled.plane
+		else if(length(grabbed_by))
+			for(var/obj/item/grab/G AS_ANYTHING in grabbed_by)
+				if(G.current_grab.adjust_plane)
+					new_plane = max(new_plane, G.assailant.plane)
+	if(last_plane != new_plane)
+		plane = new_plane
+		UPDATE_OO_IF_PRESENT
+
+/mob/living/get_base_layer()
+	if(jumping)
+		return VEHICLE_LOAD_LAYER
+	if (hiding)
+		return HIDING_MOB_LAYER
+	. = ..()
+
+/mob/living/carbon/human/get_base_layer()
+	if(lying)
+		return LYING_HUMAN_LAYER
+	. = ..()
+
+/mob/reset_offsets(var/anim_time = 2)
```

Comment:
If I understand your approach correctly, maybe worth commenting here (and above for layer) that if anything ever touches mob offsets, it must do so here and here alone, and it must call this proc to enact whatever change it desires.

I don't necessarily love that, but I can see how it's much better than stuff setting these ad-hoc.

Reply:
Alrighty, will add a comment.

Reply:
Added.